---
title: "Integrative single-cell clustering analysis"
author: "Wu"
output:
  slidy_presentation: default
  ioslides_presentation:
    transition: faster
    widescreen: yes
  beamer_presentation: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, fig.width=6, fig.height=4, fig.align = "center", warning=FALSE, message=FALSE)
```


## Introduction
The recent advances in molecular biology, microfluidics and nanotechnology open a new avenue for the precision medicine. The key to the precison treatment is discovering the heterogeneity at the cellular level. The maturation of single-cell sequencing technologies provides unique opportunities for us to detect the characteristics of individual cell. Single-cell RNA sequencing (scRNA-seq) is one of the most widely used single-cell sequencing approaches. Although scRNA-seq technologies have yielded transformative insights into cellular diversity and development, they also limit the ability to derive a deep understanding of the relationships between biomolecules in single cells. The analysis of scRNA-seq data in conjunction with other data types contributes to study cells from multiple perspectives. 

## Introduction
In this project, finding a clustering method to detect the heterogeneity between single cells and integrating scRNA-seq and DNA methylation data to make clustering more accurate are the core. DNA methylation data is a good complemet to scRNA-seq data since methylation in the vicinity of the promoter is associated with the absence of transcription. Angermueller, C. et al. performed parallel DNA methylome and transcriptome sequencing in the same cell and examines the relationships between DNA methylation and gene expression. Linker, S. M. et al. applied parallel DNA methylation and transcriptome sequencing to differentiate human induced pluripotent stem cells. Mo, Qianxing, et al. used a fully Bayesian method to model continuous and discrete omics data. But there is no efficient method which integrates multi-data for analysis.

We have developed a Bayesian integrative clustering method which takes the count nature, over-dispersion and zero-inflated effects of data into account. The new model is a valuable tool for multi-data analysis. We organize this article as follows. In Section 2, we provide details for the model. In Section 3, we show that the model can reveal meaningful subtypes by analyzing serum embryonic stem cells (ESCs) data. In Section 4, we summarize this article with a brief discussion.

## Model
Suppose that both the scRNA-seq and DNA methylation data are collected for $N$ cells in some tissue. The RNA expression levels for $G$ genes and methylated counts for $J$ covered cytosine positions are measured in each cell. We denote by $X_{gi}$ and $M_{ji}$ the read count mapped to gene $g$ and position $j$ in cell $i$, respectively. 

The core idea of the integrative clustering framework is to reduce the multi-high dimensional space to a low dimensional subspace that will collectively capture the major variations of the multiple genomic data sets. Therefore, the low-dimensional subspace can be used to cluster the cell samples. We assume a low-dimensional integrated subspace $\mathbf{Z}$ with dimension $N \times K$, where there are $K+1$ subtypes of cell samples. In other words, each sample is associated with a latent variable $\mathbf{z}_i = (z_{i1}, \cdots, z_{ik}), (i = 1, \cdots, N)$. 

## Model
For DNA methylation data, we model the data with the Poisson model
$$M_{ji} \sim \operatorname{Poisson}(e^{\mathbf{z}^*_i \Omega_j \alpha_j}), (i = 1, \cdots, N, j = 1, \cdots, J) ,$$
where $\alpha_j = \left( \alpha_{0j} , \alpha_{1j} , \cdots, \alpha_{Kj} \right)^{T}$ is the coefficient vector associated with the $j$ th possion; $\mathbf{z}^*_i = \left( 1 , \mathbf{z}_{i} \right) = \left(1 ,z_{i1}, \cdots, z_{ik} \right)$, is a vector in which the first component is 1 and the remaining components are exactly from vector $\mathbf{z}_{i}$; $\Omega_j = \operatorname {diag} \left(1 , \omega_j , \cdots , \omega_j \right)$ is a diagonal matrix whose first diagonal component is 1 and all the remaining $K$ diagonal components are $\omega_j$. In the model, $\omega_j$ is an indicator variable with value 0 or 1, which is used for Bayesian variable selection. When $\omega_j = 0$, it indicates that the corresponding $\alpha_j$ is small and thus the $j$ th possion contributes little for the joint clustering. Otherwise, the corresponding $\alpha_j$ is large and thus the $j$ th possion is a contributor for the joint clustering.

## Model
For scRNA-seq data, we take zero-inflated effects into consideration with the following model
$$Y_{gi} \sim \operatorname{Poisson}(e^{\mathbf{z}^*_i \Gamma_g \beta_g}), (i = 1, \cdots, N, g = 1, \cdots, G) ,$$
$$X_{gi} = \left\{ \begin{array} { l l } { 0 } & { \text { with probability } p(Y_{gi}) } \\ 
{Y_{gi}} & { \text { with probability } 1 - p(Y_{gi}) } \end{array} \right. ,$$
where $\beta_g = \left( \beta_{0g} , \beta_{1g} , \cdots, \beta_{Kg} \right)^{T}$ is the coefficient vector associated with the $g$ th gene; $\Gamma_g = \operatorname {diag} \left(1 , \gamma_g , \cdots , \gamma_g \right)$ is a diagonal matrix whose first diagonal component is 1 and all the remaining $K$ diagonal components are $\gamma_g$, which is an indicator variable with value 0 or 1; $Y_{gi}$ is the latent variable of the true read count mapped to gene $g$ in cell $i$; $p(Y_{gi})$ is the dropout rate defined as $\Phi \left(\lambda_{g0} + \lambda_{g1} log_2(Y_{gi}+1) \right)$, where $\Phi(·)$ is the cumulative distribution function of the standard normal distribution. The negative $\lambda _ {g1}$ controls the intensity of becoming zero and is associated with gene $g$. Because $p(Y_{gi})$ is a decreasing function, the larger the $Y_{gi},$ the less likely that we observe a zero for $X_{gi}$.

## Priors
To perform Bayesian analysis,we assume the following prior distributions for the model parameters
$$\mathbf{z}_i \stackrel{i.i.d.} { \sim } \mathrm{ MVN } \left( \mathbf{z}_{0} ,\Sigma_{\mathbf{z}} \right), (i = 1, \cdots, N), $$
where $\mathbf{z}_i$ follows a multivariate normal distribution.
$$ \alpha_j \stackrel{i.i.d.} { \sim } \mathrm{ MVN } \left( \alpha_{0} , \Sigma_{\alpha} \right), \omega_j \stackrel{i.i.d.} { \sim } \operatorname{Bernoulli}(q_\omega), (j = 1, \cdots, J),$$
$$\beta_g \stackrel{i.i.d.} { \sim } \mathrm{ MVN } \left( \beta_{0} , \Sigma_{\beta} \right), \gamma_g \stackrel{i.i.d.} { \sim } \operatorname{Bernoulli}( q_\gamma), \lambda_{g0} \stackrel{i.i.d.} { \sim } \mathrm{N}(\lambda_{0}, \sigma^2_0), \lambda_{g1} \stackrel{i.i.d.} { \sim } \mathrm{N}(\lambda_{1}, \sigma^2_1), (g = 1, \cdots, G).$$

## Posteriors
Given the priors, we utilize the blocked Gibbs sampler to carry out the posterior sampling. In each iteration, the sampling scheme proceeds as follows. ("-" means given all other variables) 

1. Sample the missing variable $Y_{gi}$ for which its observation $X_{gi}$ equals zero from 
$$ p(Y_{gi}| - ) \propto \begin{cases} 
\frac{ e^{\mathbf{z}^*_i \Gamma_g \beta_g Y_{gi}}}{Y_{gi}!} \Phi \left(\lambda_{g0} + \lambda_{g1} log_2(Y_{gi}+1) \right),  & \mbox{if }Y_{gi} > 0 \\
\frac{ e^{\mathbf{z}^*_i \Gamma_g \beta_g Y_{gi}}}{Y_{gi}!} , & \mbox{if }Y_{gi} = 0
\end{cases},$$
where $\Phi(\dot)$ is the cumulative distribution function of the standard normal distribution.
2. Update the coefficient vector $\alpha_j$ associated with the $j$ th position from
$$p( \alpha_j | - ) \propto \left( \prod_{i=1}^N \left( \exp ( \mathbf{z}^*_i \Omega_j \alpha_j) \right) ^ {M_{ji}} \exp \left( - \exp ( \mathbf{z}^*_i \Omega_j \alpha_j) \right) \right) \exp \left( - \frac{1}{2} \left( \alpha_j - \alpha_{0} \right) ^ { T } \Sigma_{\alpha}^{-1} \left( \alpha_j - \alpha_{0} \right) \right).$$
3. Update the indicator variable $\omega_j$ associated with the $j$ th position from
$$p( \omega_j | - ) \propto \left( \prod_{i=1}^N \left( \exp ( \mathbf{z}^*_i \Omega_j \alpha_j) \right) ^ {M_{ji}} \exp \left( - \exp ( \mathbf{z}^*_i \Omega_j \alpha_j) \right) \right) p(\omega_j).$$
4. Update the coefficient vector $\beta_g$ associated with the $g$ th gene from
$$p( \beta_g | - ) \propto \left( \prod_{i=1}^N \left( \exp ( \mathbf{z}^*_i \Gamma_g \beta_g) \right) ^ {Y_{gi}} \exp \left( - \exp ( \mathbf{z}^*_i \Gamma_g \beta_g) \right) \right) \exp \left( - \frac{1}{2} \left( \beta_g - \beta_{0} \right) ^ { T } \Sigma_{\beta}^{-1} \left( \beta_g - \beta_{0} \right) \right).$$
5. Update the indicator variable $\gamma_g$ associated with the $g$ th gene from
$$p( \gamma_g | - ) \propto \left( \prod_{i=1}^N \left( \exp ( \mathbf{z}^*_i \Gamma_g \beta_g) \right) ^ {Y_{gi}} \exp \left( - \exp ( \mathbf{z}^*_i \Gamma_g \beta_g) \right) \right) p(\gamma_g).$$
6. Update the zero-ination intensity parameters $\lambda_{g0}$ and $\lambda_{g1}$ by generating
$$\lambda_{g0} \sim p(\lambda_{g0}|-) \propto e^{-\frac{(\lambda_{g0}-\lambda_{0})^2}{2 \sigma^2_0}} \prod\limits_{(j,i):X^{(j)}_{gi}=0, Y^{(j)}_{gi}>0}\Phi(\lambda_{g0} + \lambda_{g1}\log_2(Y^{(j)}_{gi}+1))  \prod\limits_{(j,i):X^{(j)}_{gi}>0}(1-\Phi(\lambda_{g0} + \lambda_{g1}\log_2(Y^{(j)}_{gi}+1)))$$
$$\lambda_{g1} \sim p(\lambda_{g1}|-) \propto e^{-\frac{(\lambda_{g1}-\lambda_{1})^2}{2 \sigma^2_1}} \prod\limits_{(j,i):X^{(j)}_{gi}=0, Y^{(j)}_{gi}>0}\Phi(\lambda_{g0} + \lambda_{g1}\log_2(Y^{(j)}_{gi}+1)) \prod\limits_{(j,i):X^{(j)}_{gi}>0}(1-\Phi(\lambda_{g0} + \lambda_{g1}\log_2(Y^{(j)}_{gi}+1))).$$
7. Update the latent variable $\mathbf{z}_i$ associated with the $i$ th cell sample from 
$$p(\mathbf{z}_i | -) \propto p(\mathbf{z}_i)\left( \prod_{g=1}^{G}\left( \exp ( \mathbf{z}^*_i \Gamma_g \beta_g) \right) ^ {Y_{gi}} \exp \left( - \exp ( \mathbf{z}^*_i \Gamma_g \beta_g) \right)\right) \left(\prod_{j=1} ^ {J} \left( \exp ( \mathbf{z}^*_i \Omega_j \alpha_j) \right) ^ {M_{ji}} \exp \left( - \exp ( \mathbf{z}^*_i \Omega_j \alpha_j) \right) \right).$$

The Metropolis–Hasting algorithm will be used to sample from its posterior distribution for statistical inference (Metropolis and others, 1953; Hastings, 1970). The posterior mean value of the latent variables are used for sample clustering. Specifically, Bayesian information criterion (BIC) is applied to select cluster number $K+1$. Following a general principle for separating $K+1$ clusters among $N$ data points, we use k-means clustering to divide the $N$ samples into $K+1$ clusters in the latent variable space.

## Results
The model was performed on a real world data set including 61 mouse serum embryonic stem cells. This dataset is publicly available with the accession code GSE74535 on the website of NCBI. The data sets consists of 61 ESCs that were measured by scRNA-seq expression of 7356 genes and DNA methylation levels at more than three million positions. In the preprocessing procedure, the same DNA methylation positions on the promoters for all ESCs were selected since methylation in the vicinity of the promoter is associated with the absence of transcription. For RNA expression, 500 most variable genes were used for clustering analysis for a good cell cluster separation and the scRNA-seq data were normalized to correct for technical factors: the library size for each cell, the sum of read counts across all genes, and the median of all library sizes were calculated and the original counts were divided by its corresponding library size and multiplied the ratio by the median library size. We took the floor of normalized data. Therefore, both the 500 gene expression and 89 DNA methylation positions data were presented as counts and used in the clustering analysis.

In the integrative clustering analysis, the priors for parameters were set as following: $\mathbf{z}_{0} = \mathbf{0}_{K}, \Sigma_{\mathbf{z}} = \mathbf{I}_{K}, \alpha_{0} = \beta_{0} = \mathbf{0}_{K+1}, \Sigma_{\alpha} = \Sigma_{\beta} = \mathbf{I}_{K+1} , q_\omega = q_\gamma = 0.5, \lambda_{0} = 0, \sigma^2_0 = 1, \lambda_{1} = -1 , \sigma^2_1 = 0.01.$ The cluster number parameter $K$ from 1 to 4 needed to be tested. For each $K$, we ran 5000 MCMC iterations, of which the first 2500 were discarded as burn-in.

<div align=center>
```{r 111, echo=FALSE, message=FALSE, warning=FALSE}
setwd("../data")
load("RNA_DNAm_2.RData")
BICvalue <- c(379599.7, 392696.1, 386628.4, 393845)
Kvalue <- c(1,2,3,4)

BICdata <- as.data.frame(cbind(Kvalue, BICvalue))
library(plotly)
plot_ly(data = BICdata, x = ~as.factor(Kvalue), y = ~BICvalue, mode = 'lines+markers',type = 'scatter')%>%
  layout(
    title = "Figure 1: BIC values",
    xaxis = list(title = "K"),
    yaxis = list(title = "BIC"))

Z_est <- apply(collection_z, c(2,3), mean)
getMode <- function(x){
  return(as.numeric(names(table(x))[table(x) == max(table(x))][1]))
}

Gam_est <- apply(collection_Gam, 2, getMode)
Omega_est <- apply(collection_Omega, 2, getMode)

library(factoextra)
set.seed(20190330)
km_result <- kmeans(Z_est, 2)
d1 <- c(sum(Gam_est), 500-sum(Gam_est))
d2 <- c(1,0)
Gam <- cbind(d1,d2)
p1<-plot_ly(data = as.data.frame(Gam), y = ~d1, x= ~as.factor(d2),
        colorscale = "Portland", type = 'bar', name = "Gamma"
        # ,symbol =as.factor(cell1$cell_t1),
        # symbols = c(1:10)
)%>%
  layout(
    # title = "Subject cluster 1",
    xaxis = list(title = "Gamma indicator"),
    yaxis = list(title = ""))
d1 <- c(sum(Omega_est), 89-sum(Omega_est))
d2 <- c(1,0)
Omega <- cbind(d1,d2)
p2<-plot_ly(data = as.data.frame(Omega), y = ~d1, x= ~as.factor(d2),
            colorscale = "Portland", type = 'bar', name = "Omega"
            # ,symbol =as.factor(cell1$cell_t1),
            # symbols = c(1:10)
)%>%
  layout(
    # title = "Subject cluster 1",
    xaxis = list(title = "Omega indicator"),
    yaxis = list(title = ""))
subplot(p1,p2)%>%
  layout(
    title = "Figure 2: The posterior mode of the genomic feature indicators",
    xaxis = list(title = "Indicator"),
    xaxis2 = list(title = "Indicator"))

cell <- as.data.frame(cbind(c(1:61), Z_est[,2],km_result$cluster))
plot_ly(data = cell, x = ~cell[,1], y = ~cell[,2], symbol =as.factor(cell$V3),
        colorscale = "Portland",
        mode = 'markers',type = 'scatter',showlegend=T
        # ,symbol =as.factor(cell1$cell_t1),
        # symbols = c(1:10)
)%>%
  layout(
    title = "Figure 3: The clustering result",
    xaxis = list(title = "Cell ID"),
    yaxis = list(title = "Z value"))

library(Rtsne)
set.seed(20190330)
tsne <- Rtsne(Z_est, dims = 2, perplexity=15, max_iter = 500)
cell <- as.data.frame(cbind(tsne$Y,km_result$cluster))
plot_ly(data = cell, x = ~cell[,1], y = ~cell[,2], symbol =as.factor(cell$V3),
        colorscale = "Portland",
        mode = 'markers',type = 'scatter',showlegend=T
        # ,symbol =as.factor(cell1$cell_t1),
        # symbols = c(1:10)
)%>%
  layout(
    title = "Figure 4: tSNEs of the clustering result ",
    xaxis = list(title = "tSNE1"),
    yaxis = list(title = "tSNE2"))

```
</div>


Figure 1 shows the Bayesian information criterion (BIC) values for the selected $K$. The BIC value reaches the minimum when $K$ is equal to 1, which indicates the optimal number of clusters for the samples is 2. Figure 2 shows results of the posterior mode of the genomic feature indicators. A posterior mode equal to 1 suggests that the corresponding genomic feature is more likely to be an informative feature, which contributes to the integrative clustering. We used k-means method to cluster the ESCs and Figure 3 shows the clustering result in the latent varaible space. Obviously, we can see that two clusters are very separate, indicating there indeed exists heterogeneity among ESCs.

The differentially expressed (DE) genes for ESCs were detected by the posterior mode of indicator variable $\gamma_g (g = 1, \cdots, G).$ Regarding DE genes between two sample clusters, we found that the most significant pathways are related to the embryonic development, tumor growth, uptake of glucose, bone remodeling and immune functions. The two types of ESCs would have different directions of differentiation.

## Discussion
We develop a Bayesian model to discover cell heterogeneity by integrating scRNA-seq and DNA methylation data. The dropout rate and Poisson distribution employed in the model can account for the count nature, over-dispersion and zero-inflated effects in the data. Our model is flexible and computationally efficient for integrative analysis of multi-data. Especially, the model uses a latent space to bridge the methodology gap between two data types. The framework in our model can be further adapted to any situations where the observed data is sparse, count-valued and for clustering. With more multi-data generated by advanced technology, our model can be a useful tool to facilitate the development of the personalized treatment.

## References
Angermueller, Christof, et al. "Parallel single-cell sequencing links transcriptional and epigenetic heterogeneity." Nature methods 13.3 (2016): 229.

Linker, Stephanie M., et al. "Combined single-cell profiling of expression and DNA methylation reveals splicing regulation and heterogeneity." Genome Biology 20.1 (2019): 30.

Mo, Qianxing, et al. "A fully Bayesian latent variable model for integrative clustering analysis of multi-type omics data." Biostatistics 19.1 (2017): 71-86.
 
George, Edward I., and Robert E. McCulloch. "Approaches for Bayesian variable selection." Statistica sinica (1997): 339-373.